WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift

```diff
@@ -273,46 +281,58 @@ private extension DashboardViewModel {
+        // When not available, Analytics cards need to be hidden from Dashboard, but appear on Customize as "Unavailable"
+        cards.append(DashboardCard(type: .performance,
+                                   availability: canShowAnalytics ? .show : .unavailable,
+                                   enabled: canShowAnalytics))
+
+        cards.append(DashboardCard(type: .topPerformers,
+                                   availability: canShowAnalytics ? .show : .unavailable,
+                                   enabled: canShowAnalytics))
+
+        // Blaze card.
+        // When not available, Blaze card needs to be hidden from Dashboard and Customize
+        cards.append(DashboardCard(type: .blaze,
+                                   availability: canShowBlaze ? .show : .hide,
+                                   enabled: canShowBlaze))
+
+        return cards
+    }
+
     @MainActor
     func updateDashboardCards(canShowOnboarding: Bool,
                               canShowBlaze: Bool,
                               canShowAnalytics: Bool) async {
-        dashboardCards = await {
-            if var stored = await loadDashboardCards() {
-                let analyticCardTypes: [DashboardCard.CardType] = [.performance, .topPerformers]
-                stored = canShowAnalytics ? stored : stored.filter { !analyticCardTypes.contains($0.type) }
-                return stored
+
+        // First, generate latest cards state based on current canShow states
+        let initialCards = generateDefaultCards(canShowOnboarding: canShowOnboarding,
+                                                canShowBlaze: canShowBlaze,
+                                                canShowAnalytics: canShowAnalytics)
+
+        // Next, get saved cards and preserve existing enabled state for all available cards.
+        // This is needed because even if a user already disabled an available card and saved it, in `initialCards`
+        // the same card might be set to be enabled. To respect user's setting, we need to check the saved state and re-apply it.
+        let savedCards = await loadDashboardCards() ?? []
+        dashboardCards = initialCards.map { initialCard in
+            if let savedCard = savedCards.first(where: { $0.type == initialCard.type }),
+               savedCard.availability == .show && initialCard.availability == .show {
+                return initialCard.copy(enabled: savedCard.enabled)
             } else {
-                // Start with default values, cards could be updated further below as needed.
-                return [DashboardCard(type: .onboarding, enabled: canShowOnboarding),
-                        DashboardCard(type: .performance, enabled: canShowAnalytics),
-                        DashboardCard(type: .topPerformers, enabled: canShowAnalytics),
-                        DashboardCard(type: .blaze, enabled: canShowBlaze)]
+                return initialCard
```

Comment:
In this function, the logic for updating is:
1. Create set of default cards based on `canShow` parameters,
2. Compare those default cards with the saved cards, and take into account saved card's enabled value if the same card in default and saved are both `.show`

The comparison is needed to deal with this example case:

1. Imagine initially the site has 2 available cards: `onboarding`, `performance`
2. Then, the user sets `onboarding` as enabled, `performance` as disabled, and saves it,
3. Then, the user sets up the site so that Blaze is now available for the site,
4. That triggers `canShowBlaze` value update, which then calls `updateDashboardCards()`
5. Inside the function, `generateDefaultCards()` is then called to create `initialCards`
6. In that array, `performance` is set as enabled (because `canShowAnalytics` remains true as before)
7. Now, we can't set `dashboardCards` with `initialCards` right away, because from point 2 above we know that the user wants `performance` to be disabled.
8. Thus, we need to compare `initialCards` with `savedCards`, find card that's both `.show` in the arrays (indicating they're unchanged after card update), then take the `enabled` state from the saved card.

Reply:
Thanks for the clear explanation, this makes sense!